dali/python/nvidia/dali/pipeline.py

```diff
@@ -990,3 +996,93 @@ def iter_setup(self):
         For example, one can use this function to feed the input
         data from NumPy arrays."""
         pass
+
+
+def _discriminate_args(func, **func_kwargs):
+    """Split args on those applicable to Pipeline constructor and the decorated function."""
+    func_argspec = inspect.getfullargspec(func)
+    ctor_argspec = inspect.getfullargspec(Pipeline.__init__)
+
+    ctor_args = {}
+    fn_args = {}
+
+    for farg in func_kwargs.items():
+        if farg[0] in func_argspec.args or farg[0] in func_argspec.kwonlyargs:
+            fn_args[farg[0]] = farg[1]
+        elif farg[0] in ctor_argspec.args or farg[0] in ctor_argspec.kwonlyargs:
+            ctor_args[farg[0]] = farg[1]
+        else:
+            fn_args[farg[0]] = farg[1]
+
+    return ctor_args, fn_args
+
+
+def pipeline_def(fn=None, **pipeline_kwargs):
+    """
+    Decorator that converts a graph definition function into a DALI pipeline factory.
+
+    A graph definition function is a function that returns intended pipeline outputs.
+    You can decorate this function with ``@pipeline_def``::
+
+        @pipeline_def
+        def my_pipe(flip_vertical, flip_horizontal):
+            ''' Creates a DALI pipeline, which returns flipped and original images '''
+            data, _ = fn.file_reader(file_root=images_dir)
+            img = fn.image_decoder(data, device="mixed")
+            flipped = fn.flip(img, horizontal=flip_horizontal, vertical=flip_vertical)
+            return flipped, img
+
+    The decorated function returns a DALI Pipeline object::
+
+        pipe = my_pipe(True, False)
+        # pipe.build()  # the pipeline is not configured properly yet
+
+    A pipeline requires additional parameters such as batch size, number of worker threads,
+    GPU device id and so on (see :meth:`Pipeline.__init__` for a complete list of pipeline parameters).
+    These parameters can be supplied as additional keyword arguments, passed to the decorated function::
+
+        pipe = my_pipe(True, False, batch_size=32, num_threads=1, device_id=0)
+        pipe.build()  # the pipeline is properly configured, we can build it now
+
+    The outputs from the original function became the outputs of the Pipeline::
+
+        flipped, img = pipe.run()
+
+    When some of the pipeline parameters are fixed, they can be specified by name in the decorator::
+
+        @pipeline_def(batch_size=42, num_threads=3)
+        def my_pipe(flip_vertical, flip_horizontal):
+            ...
+
+    Any Pipeline constructor parameter passed later when calling the decorated function will
+    override the decorator-defined params::
+
+        @pipeline_def(batch_size=32, num_threads=3)
+        def my_pipe():
+            data = fn.external_source(source=my_generator)
+            return data
+
+        pipe = my_pipe(batch_size=128)  # batch_size=128 overrides batch_size=32
+
+    .. warning::
```

Comment:
```suggestion
    .. warning::
        Avoid the use of `**kwargs` in the graph definition function, since it may result in unwanted,
        silent hijacking of some arguments of the same name by Pipeline constructor. Code written
        this way may cease to work with future versions of DALI when new parameters are added
        to the Pipeline constructor.
        
    .. warning::
```

Reply:
Done

Reply:
Can't we just disallow the `**kwargs` completely for the function that is to be decorated and raise an error. I'd say play safe rather than write lengthy manuals. If we find a better way or this is an issue we can always relax that check (and doing the other way would be hard).